### PR TITLE
feat: show all tokens for all chains in the token list

### PR DESCRIPTION
**Pull request by jonybur**
## **Description**

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27785?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by darkwing** [comment]
Thank you for posting this WIP Jony.  One approach that may make this easier is starting with storybook and adding an AssetList story with mock data across all chains and ensuring it looks good there.  That will likely make the rest easier.

**Comment by gambinish** [comment]
Hey @jonybur nice work getting this started! I requested myself as review here since we'll be working in the same area. I'll request you on my PRs shortly. I know this is still pretty early on, but a couple things I noticed to put on your radar. You are probably already investigating these things!

We'll want to make sure that we include the native tokens in the Sort by function to avoid a regression there:

https://github.com/user-attachments/assets/5e20d995-9fb8-4f85-831f-e8f547c164de

Does the `TokenBalanceController` need to get updated with balances across networks, in order to display the balances correctly for each token, on each network? From what I can tell, I think the controller currently accepts a single `chainId` in it's constructor (the selectedNetwork). I think that ideally, it should accept an array of `chainId` so that we can have a mapping of chainId => tokenBalances for each chain. If I am missing something here, please lmk!

Additionally, we'll also need to consider how we're handling the native token balances for each chain, since that's being handled a bit differently than the erc20 tokens. Currently when I'm on Ethereum mainnet, it's calculating the POL balance at the exchange rate of ETH on mainnet, rather than POL on Polygon (and vice versa). I think this also happens at the `CurrencyRateController`

https://github.com/user-attachments/assets/6241d358-2b75-4b26-bb9a-16d699d8acc9

We should also take into consideration the `showZeroBalances` flag if you haven't already 👍

File: ui/selectors/multichain.ts [hunk 1]
```diff
@@ -184,6 +185,110 @@ export function getMultichainNetwork(
...
+  return chainId;
+}
+
+export function hexToCaipChainId(hex: string): CaipChainId {
+  const hexValue = hex.startsWith('0x') ? hex.slice(2) : hex;
+  const chainIdDecimal = parseInt(hexValue, 16);
+  return `eip155:${chainIdDecimal}`;
+}
+
+export function getChains(state: MultichainState): MultichainNetwork[] {
```
Comment (bergeron) [inline ui/selectors/multichain.ts:203]: Whats the reason for the changes in this file?  I understood it was for non-evm support, so wondering why it needs to change vs just having a selector in `selectors.js`

**Comment by salimtb** [comment]
Hi @jonybur ,

If I understand correctly, the process in this PR works like this:

- You retrieve a list of all tokens along with their balances using `accountsByChainId`, which provides a list of token addresses and balances, organized by blockchain network (so this results in a very large amount of data).
- 
- Then, you filter the data to keep only the tokens that have a balance. This way, you get the list of "detected tokens" across all chains by looping through them (link to code [here](https://github.com/MetaMask/metamask-extension/pull/27785/files#diff-45ae84764c1d49a07813a91adc3351d6ed96bf9ee8fa2ae4c9a116d1e0cba06cR455-R462)).
- After that, you display the converted values.

This approach has a significant impact on performance because it handles a large dataset. To improve this, I suggest simplifying it by making the marketData multichain in the assetsController. The marketData already includes the detected tokens, tokens imported by the user, and all market-related information, but right now, it only handles data for the selected network.
you can fin below the structure of market data:
```
{
  "0x1": {
    // Native token here ******
    "0x0000000000000000000000000000000000000000": {
      "tokenAddress": "0x0000000000000000000000000000000000000000",
      "currency": "ETH",
      "id": "ethereum",
      "price": 1.0000726178421582,
      "marketCap": 120374263.85039304,
      "allTimeHigh": 1.872695124126123,
      "allTimeLow": 0.00016621452365167183,
      "totalVolume": 5308887.015842034,
      "high1d": 1.0147677692591692,
      "low1d": 0.9936194962298293,
      "circulatingSupply": 120387362.762955,
      "dilutedMarketCap": 120374263.85039304,
      "marketCapPercentChange1d": -0.64964,
      "priceChange1d": -16.22303390018078,
      "pricePercentChange1h": -0.2999715552409102,
      "pricePercentChange1d": -0.618879494740698,
      "pricePercentChange7d": 9.728879673482494,
      "pricePercentChange14d": 12.175196996309179,
      "pricePercentChange30d": 10.175859028915708,
      "pricePercentChange200d": -28.044745794161667,
      "pricePercentChange1y": 64.93607348906181
    },
    // Detected/imported token here ******
    "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48": {
      "tokenAddress": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
      "currency": "ETH",
      "id": "usd-coin",
      "price": 0.00038370084519404344,
      "marketCap": 13331889.134607311,
      "allTimeHigh": 0.00044914647747917576,
      "allTimeLow": 0.00033691628933347535,
      "totalVolume": 1574846.8115542484,
      "high1d": 0.0003850375358218917,
      "low1d": 0.00038302674159194645,
      "circulatingSupply": 34751452795.3628,
      "dilutedMarketCap": 13331829.76013019,
      "marketCapPercentChange1d": -0.06737,
      "priceChange1d": -0.000655467524627418,
      "pricePercentChange1h": -0.05174268708087084,
      "pricePercentChange1d": -0.06553537070828762,
      "pricePercentChange7d": 0.02156004626949369,
      "pricePercentChange14d": -0.03760749615660747,
      "pricePercentChange30d": 0.01603305301773797,
      "pricePercentChange200d": 0.05198102213414082,
      "pricePercentChange1y": -0.012088418136522984
    },
  }
}
```
In the meantime, I recommend using mock data until we can handle this properly on the backend (specifically in the assetsController).

**Comment by bergeron** [comment]
> You retrieve a list of all tokens along with their balances using accountsByChainId, which provides a list of token addresses and balances

I believe `accountsByChainId` only stores native balances, so it would not deal with tokens.  It's from the `AccountTrackerController` and is a mapping from chain id -> account address -> native balance

File: ui/components/app/assets/token-list/token-list.tsx [hunk 2]
```diff
@@ -52,10 +112,7 @@ export default function TokenList({
   };
 
   const sortedTokens = useMemo(() => {
-    return sortAssets(
-      [nativeTokenWithBalance, ...tokensWithBalances],
-      tokenSortConfig,
-    );
```
Comment (gambinish) [inline ui/components/app/assets/token-list/token-list.tsx:58]: Why are we removing native tokens from the sorting functionality here? We will need to include the native tokens with their balances, along with all erc20's so that they can be sorted in the correct order by declining fiat balance.

File: ui/components/multichain/asset-picker-amount/asset-picker-modal/AssetList.tsx [hunk 3]
```diff
@@ -139,3 +128,56 @@ export default function AssetList({
...
+      chain={chain}
+      primary={primaryCurrencyProperties.value}
+      tokenSymbol={primaryCurrency}
+      secondary={secondaryCurrencyDisplay}
+      tokenImage={token.image}
+      isOriginalTokenSymbol
+      tooltipText={isDisabled ? 'swapTokenNotAvailable' : undefined}
+    />
+  );
+}
```
Comment (gambinish) [inline ui/components/multichain/asset-picker-amount/asset-picker-modal/AssetList.tsx:183]: Is there a reason why these changes can't be incorporated into the `TokenListItem` component? This seems to essentially wrap that component.

File: ui/components/app/assets/token-list/token-list.tsx [hunk 4]
```diff
@@ -33,17 +37,73 @@ export default function TokenList({
...
+          key={chain.network.chainId}
+          chain={chain}
+          onTokenClick={onTokenClick}
+          tokenSortConfig={tokenSortConfig}
+          selectedAccount={selectedAccount}
+          conversionRate={conversionRate}
+          contractExchangeRates={contractExchangeRates}
+          shouldHideZeroBalanceTokens={shouldHideZeroBalanceTokens}
+        />
+      ))}
```
Comment (gambinish) [inline ui/components/app/assets/token-list/token-list.tsx:77]: Rather than mapping over chains and rendering a separate list for each chain and pinning the `nativeToken` to the top, I would suggest doing some data manipulation first. I think it would make sense to move the sortedAssets `useMemo` function up to this level, before the return.

This will allow us to:

1. Make this into a flat array first, so that we don't have a separate `TokenList` for each chain. We should instead have a single `TokenList`, which is correctly sorted and filtered for all tokens for all chains

2. Allow us to include the native tokens for each chain in the sorted/filtered array, rather than pinning them as a fragment to the top as we are here.

File: ui/components/app/assets/token-list/token-list.tsx [hunk 5]
```diff
@@ -76,17 +133,12 @@ export default function TokenList({
   ) : (
     <div>
       {sortedTokens.map((tokenData) => {
-        if (tokenData?.isNative) {
-          // we need cloneElement so that we can pass the unique key
-          return React.cloneElement(nativeToken as React.ReactElement, {
-            key: `${tokenData.symbol}-${tokenData.address}`,
-          });
-        }
```
Comment (gambinish) [inline ui/components/app/assets/token-list/token-list.tsx:84]: I think that if we include the `nativeTokens` as an array of react nodes in `TokenList`, rather than a fragment as we are now, we can retain this logic here.

The idea: when all tokens (including native tokens) get sorted and filtered together, each native token can be identified by the `isNative` prop. React can then render them as a `NativeToken` component, while still retaining it's sorted order within the list.

File: ui/components/app/assets/asset-list/asset-list.tsx [hunk 6]
```diff
@@ -115,7 +118,13 @@ const AssetList = ({ onClickAsset, showTokensLinks }: AssetListProps) => {
...
       <AssetListControlBar showTokensLinks={showTokensLinks} />
       <TokenList
-        nativeToken={<NativeToken onClickAsset={onClickAsset} />}
+        nativeToken={
+          <>
+            {chains?.map((chain) => (
+              <NativeToken chain={chain} onClickAsset={onClickAsset} />
+            ))}
+          </>
+        }
```
Comment (gambinish) [inline ui/components/app/assets/asset-list/asset-list.tsx:127]: Let's rename this prop to `nativeTokens`

Rather than wrapping them in a fragment, let's just pass them in as an array of `NativeToken` elements
```suggestion
        nativeToken={chains?.map((chain) => (
          <NativeToken chain={chain} onClickAsset={onClickAsset} />
        ))}
```

This will allow us to render each of them independently farther down the component tree, since they won't be bundled together in a fragment. They will need to appear in different orders in the list depending on sort and filter criteria

**Review by gambinish — CHANGES_REQUESTED** [review]
Hey Jony, went over this again and I left a few suggestions mainly related to how we can avoid regressions to our sorting functionality. This sortedAssets `useMemo` looks like it will be our main integration point for our features, and I plan to incorporate `filteredAssets` there.

If we can do the data manipulation to covert all tokens for all chains into a flat array within/before we hit that `useMemo` hook, I think it should be a pretty straightforward integration.

I'd really like to see that hook get used higher up in the tree, so that it only gets called once, while keeping the `nativeTokenBalances` in its calculation to maintain it's behavior.

**Comment by gambinish** [comment]
As an aside, I synced with @bergeron and @salimtb and I think that there are indeed some changes that we should make in the `TokenRatesController` to provide you with `marketData` for all chains (rather than only the selected chain). This should unblock and allow you to make the balance calculations for each multichain token as you'd need to.

File: ui/components/app/assets/asset-list/asset-list.tsx (hunk 6 above)
Comment (gambinish) [reply ui/components/app/assets/asset-list/asset-list.tsx:127]: Also, if you see an alternate approach to solve for this problem, I'm open to discussing that as well 👍

File: ui/components/app/assets/token-list/token-list.tsx (hunk 2 above)
Comment (jonybur) [reply ui/components/app/assets/token-list/token-list.tsx:58]: This is a wip, I removed it to simplify my testing. I'm looking at hardcoding marketData now

File: ui/selectors/multichain.ts (hunk 1 above)
Comment (jonybur) [reply ui/selectors/multichain.ts:203]: I moved this to selectors.js

Comment (jonybur) [reply ui/selectors/multichain.ts:203]: I moved this to selectors.js

**Comment by gambinish** [comment]
Closing this in favor of https://github.com/MetaMask/metamask-extension/pull/28386
